src/editors/playgroundController.ts

```diff
@@ -492,33 +492,31 @@ export default class PlaygroundController {
     }
 
     try {
-      const progressResult = await vscode.window.withProgress(
+      return await vscode.window.withProgress(
         {
           location: ProgressLocation.Notification,
           title: 'Running MongoDB playground...',
           cancellable: true,
         },
-        async (progress, token) => {
+        async (progress, token): Promise<ShellEvaluateResult> => {
           token.onCancellationRequested(() => {
-            // If a user clicked the cancel button terminate all playground scripts.
+            // If the user clicks the cancel button,
+            // terminate all processes running on the language server.
             this._languageServerController.cancelAll();
-
-            return { result: undefined };
           });
 
-          // Run all playground scripts.
-          const result: ShellEvaluateResult = await this._evaluate(
-            codeToEvaluate
-          );
-
-          return result;
+          return Promise.race([
+            this._evaluate(codeToEvaluate),
+            new Promise<{ result: undefined }>((resolve) =>
+              token.onCancellationRequested(() => {
+                resolve({ result: undefined });
+              })
```

Comment:
Rather than racing the promise here, a better approach would be to pass in the cancellation token to `_evaluate`.  As far as I can tell we already support cancellation in `LanguageServerController`, so it's a matter of threading the token through a handful of methods until it gets there.